File: 6. Zigzag Conversion Medium/solution.go
package main

import (
	"fmt"
)

func main() {
	fmt.Println(convert("PAYPALISHIRING", 3))
	fmt.Println(convert("PAYPALISHIRING", 4))
	fmt.Println(convert("A", 1))
}

func convert(s string, numRows int) string {
	if numRows == 1 || numRows >= len(s) {
		return s
	}

	result := make([]byte, len(s))
	rowSize := 2*numRows - 2
	index := 0

	for i := 0; i < numRows; i++ {
		for j := i; j < len(s); j += rowSize {
			result[index] = s[j]
			index++
			if i != 0 && i != numRows-1 {
				diagonal := j + rowSize - 2*i
				if diagonal < len(s) {
					result[index] = s[diagonal]
					index++
				}
			}
		}
	}

	return string(result)
}

File: 6. Zigzag Conversion Medium/solution.java
class Solution {
    public static void main(String[] args) {
        System.out.println(convert("PAYPALISHIRING", 3));
        System.out.println(convert("PAYPALISHIRING", 4));
        System.out.println(convert("A", 1));
    }

    public static String convert(String s, int numRows) {
        if (numRows >= s.length() || numRows == 1) {
            return s;
        }

        char[] result = new char[s.length()];
        int rowSize = 2 * numRows - 2;
        int index = 0;

        for (int i = 0; i < numRows; i++) {
            for (int j = i; j < s.length(); j += rowSize) {
                result[index++] = s.charAt(j);
                if (i != 0 && i != numRows - 1) {
                    int diagonal = j + rowSize - 2 * i;
                    if (diagonal < s.length()) {
                        result[index++] = s.charAt(diagonal);
                    }
                }
            }
        }
        return new String(result);
    }
}